### ISSUE TITLE
Unnoticed failure

### ISSUE THREAD

**Issue by crackcomm** [fission/fission#146]
`kubectl logs --namespace fission router-2621354073-g6n3p`

```
2017/03/09 04:58:57 Starting router at port 8888
2017/03/09 05:01:58 Not cached, getting new service for {hello 81870f1a-f2d6-4106-8136-8ced70255a39}
2017/03/09 05:02:32 Failed to get service for function (hello,81870f1a-f2d6-4106-8136-8ced70255a39): (Error 0) 500 Internal Server Error
172.17.0.1 - - [09/Mar/2017:05:01:58 +0000] "GET /hello HTTP/1.1" 500 32
2017/03/09 05:02:39 Not cached, getting new service for {hello 81870f1a-f2d6-4106-8136-8ced70255a39}
172.17.0.1 - - [09/Mar/2017:05:02:39 +0000] "GET /hello HTTP/1.1" 500 32
2017/03/09 05:04:30 Failed to get service for function (hello,81870f1a-f2d6-4106-8136-8ced70255a39): (Error 0) 500 Internal Server Error
2017/03/09 05:05:22 Not cached, getting new service for {hello 81870f1a-f2d6-4106-8136-8ced70255a39}
172.17.0.1 - - [09/Mar/2017:05:05:22 +0000] "GET /hello HTTP/1.1" 500 32
2017/03/09 05:14:45 Failed to get service for function (hello,81870f1a-f2d6-4106-8136-8ced70255a39): (Error 0) 500 Internal Server Error
2017/03/09 05:14:58 Not cached, getting new service for {hello 81870f1a-f2d6-4106-8136-8ced70255a39}
```

`kubectl logs --namespace fission poolmgr-3531518326-ffqpz`

```
2017/03/09 04:57:26 starting poolmgr at port 8888
2017/03/09 04:58:16 Creating pool for environment {nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}
2017/03/09 04:58:16 [{nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}] Deployment created, waiting for a ready pod
2017/03/09 05:00:35 err: Get https://10.0.0.1:443/apis/extensions/v1beta1/namespaces/fission-function/deployments/nodejs-b871ac2c-7c8e-4a07-97f7-1fcfb18960e8-htnqpfnq: unexpected EOF
2017/03/09 05:00:35 eager-create pool failed: Get https://10.0.0.1:443/apis/extensions/v1beta1/namespaces/fission-function/deployments/nodejs-b871ac2c-7c8e-4a07-97f7-1fcfb18960e8-htnqpfnq: unexpected EOF
2017/03/09 05:00:37 Creating pool for environment {nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}
2017/03/09 05:00:37 eager-create pool failed: Post https://10.0.0.1:443/apis/extensions/v1beta1/namespaces/fission-function/deployments: dial tcp 10.0.0.1:443: getsockopt: connection refused
2017/03/09 05:00:59 Creating pool for environment {nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}
2017/03/09 05:01:58 [hello] Checking for cached function service
2017/03/09 05:01:58 [hello] No cached function service found, creating one
2017/03/09 05:01:58 [hello] getting environment for function
2017/03/09 05:01:58 [&{hello 81870f1a-f2d6-4106-8136-8ced70255a39}] getting function from controller
2017/03/09 05:01:58 [&{hello 81870f1a-f2d6-4106-8136-8ced70255a39}] getting env from controller
2017/03/09 05:01:58 [hello] getting generic pool for env
2017/03/09 05:02:32 Creating pool for environment {nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}
2017/03/09 05:02:32 eager-create pool failed: Post https://10.0.0.1:443/apis/extensions/v1beta1/namespaces/fission-function/deployments: unexpected EOF
2017/03/09 05:02:32 Error: 500: Post https://10.0.0.1:443/apis/extensions/v1beta1/namespaces/fission-function/deployments: dial tcp 10.0.0.1:443: getsockopt: connection refused
2017/03/09 05:02:38 Creating pool for environment {nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}
2017/03/09 05:02:38 [{nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}] Deployment created, waiting for a ready pod
2017/03/09 05:02:39 [hello] Checking for cached function service
2017/03/09 05:02:39 [hello] No cached function service found, creating one
2017/03/09 05:02:39 [hello] getting environment for function
2017/03/09 05:02:39 [hello] getting generic pool for env
172.17.0.5 - - [09/Mar/2017:05:02:39 +0000] "POST /v1/getServiceForFunction HTTP/1.1" 500 163
2017/03/09 05:04:30 err: Get https://10.0.0.1:443/apis/extensions/v1beta1/namespaces/fission-function/deployments/nodejs-b871ac2c-7c8e-4a07-97f7-1fcfb18960e8-tuobnktl: unexpected EOF
2017/03/09 05:04:30 Creating pool for environment {nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}
2017/03/09 05:04:30 eager-create pool failed: Get https://10.0.0.1:443/apis/extensions/v1beta1/namespaces/fission-function/deployments/nodejs-b871ac2c-7c8e-4a07-97f7-1fcfb18960e8-tuobnktl: unexpected EOF
2017/03/09 05:04:30 Error: 500: Post https://10.0.0.1:443/apis/extensions/v1beta1/namespaces/fission-function/deployments: read tcp 172.17.0.6:48972->10.0.0.1:443: read: connection reset by peer
2017/03/09 05:04:32 Creating pool for environment {nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}
2017/03/09 05:04:32 eager-create pool failed: Post https://10.0.0.1:443/apis/extensions/v1beta1/namespaces/fission-function/deployments: dial tcp 10.0.0.1:443: getsockopt: connection refused
2017/03/09 05:04:44 Creating pool for environment {nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}
2017/03/09 05:04:45 [{nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}] Deployment created, waiting for a ready pod
2017/03/09 05:05:22 [hello] Checking for cached function service
2017/03/09 05:05:22 [hello] No cached function service found, creating one
2017/03/09 05:05:22 [hello] getting environment for function
2017/03/09 05:05:22 [&{hello 81870f1a-f2d6-4106-8136-8ced70255a39}] getting function from controller
2017/03/09 05:05:22 [&{hello 81870f1a-f2d6-4106-8136-8ced70255a39}] getting env from controller
2017/03/09 05:05:22 [hello] getting generic pool for env
2017/03/09 05:09:45 Creating pool for environment {nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}
2017/03/09 05:09:45 eager-create pool failed: timeout: waited too long for pod to be ready
2017/03/09 05:09:45 [{nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}] Deployment created, waiting for a ready pod
172.17.0.5 - - [09/Mar/2017:05:05:22 +0000] "POST /v1/getServiceForFunction HTTP/1.1" 500 45
2017/03/09 05:14:45 Error: 500: timeout: waited too long for pod to be ready
2017/03/09 05:14:47 Creating pool for environment {nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}
2017/03/09 05:14:47 [{nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}] Deployment created, waiting for a ready pod
2017/03/09 05:14:58 [hello] Checking for cached function service
2017/03/09 05:14:58 [hello] No cached function service found, creating one
2017/03/09 05:14:58 [hello] getting environment for function
2017/03/09 05:14:58 [&{hello 81870f1a-f2d6-4106-8136-8ced70255a39}] getting function from controller
2017/03/09 05:14:58 [&{hello 81870f1a-f2d6-4106-8136-8ced70255a39}] getting env from controller
2017/03/09 05:14:58 [hello] getting generic pool for env
2017/03/09 05:19:48 Creating pool for environment {nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}
2017/03/09 05:19:48 eager-create pool failed: timeout: waited too long for pod to be ready
2017/03/09 05:19:48 [{nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}] Deployment created, waiting for a ready pod
```

```
time curl http://$FISSION_ROUTER/hello
Internal server error (fission)
curl http://$FISSION_ROUTER/hello  0.01s user 0.01s system 0% cpu 36.947 total
```

`kubectl version`

```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.4", GitCommit:"7243c69eb523aa4377bce883e7c0dd76b84709a1", GitTreeState:"clean", BuildDate:"2017-03-07T23:53:09Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.3", GitCommit:"029c3a408176b55c30846f0faedf56aae5992e9b", GitTreeState:"clean", BuildDate:"1970-01-01T00:00:00Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"linux/amd64"}
```

`minikube version`

```
minikube version: v0.17.1
```

**Comment by yqf3139** [comment]
Looks like the kubernetes service is not reachable from fission?
```
2017/03/09 05:00:37 eager-create pool failed: Post https://10.0.0.1:443/apis/extensions/v1beta1/namespaces/fission-function/deployments: dial tcp 10.0.0.1:443: getsockopt: connection refused
```
Have a try: `kubectl --namespace=default get service` to see the service is working.

**Comment by crackcomm** [comment]
I assumed it was a fission component it was trying to connect but I reported this issue not because of it but mainly because of that:

```
2017/03/09 05:09:45 eager-create pool failed: timeout: waited too long for pod to be ready
2017/03/09 05:09:45 [{nodejs b871ac2c-7c8e-4a07-97f7-1fcfb18960e8}] Deployment created, waiting for a ready pod
```

I have re-deployed the fission and now it is okay, because the `kubernetes` service is up but it's still an issue or maybe I just don't understand error handling in fission, anyway thanks.

**Comment by yqf3139** [comment]
Yes. The error handling is not properly, the router should return a error message with details.

**Comment by soamvasani** [comment]
> 2017/03/09 05:00:37 eager-create pool failed: Post https://10.0.0.1:443/apis/extensions/v1beta1/namespaces/fission-function/deployments: dial tcp 10.0.0.1:443: getsockopt: connection refused

> 2017/03/09 05:04:30 Error: 500: Post https://10.0.0.1:443/apis/extensions/v1beta1/namespaces/fission-function/deployments: read tcp 172.17.0.6:48972->10.0.0.1:443: read: connection reset by peer

This would explain why the pool could not be created, and why you couldn't run the function.  I'm not sure why it happened.  The kubernetes apiserver had to have been up, otherwise you couldn't have installed fission.  Did the apiserver restart for some reason?  (Were you upgrading Kubernetes at the same time?)  Can you post apiserver logs from around this timeframe if you still have them?

BTW, we don't want to expose details of fission internals to our user's users, so the 500 HTTP error returned by the function is working as designed.

**Comment by crackcomm** [comment]
I don't have any logs. I used `minikube start` [then](https://github.com/fission/fission#get-and-run-fission-minikube-or-local-cluster) [demo](https://github.com/fission/fission#usage).

Sadly I cannot provide more info about what happened behind the scenes because fission is a first project that I deployed on a kubernetes, after failing I have stopped minikube and then started again after @yqf3139 suggestion and it was already ok.

I can confirm that I've got 4GB RAM and after starting minikube and fission my laptop was getting laggy, swapping RAM, so there might be some reasons why kubernetes apiserver could have failed but I am not sure.
